### DISCUSSION TITLE
What is the correct tag for abi3 universal wheels using the py-limited-api?

### DISCUSSION THREAD

**Discussion by szabolcsdombi** [pyodide/pyodide#4377] [Q&A] [answered]
Hello,

I have recently changed my module to support the python limited api.
To ship proper universal wheels I had to change the tag generated by the bdist_wheel command.
Please note that, despite it might not look correct, this is the recommended way [link1](https://github.com/pypa/cibuildwheel/pull/1091) [link2](https://github.com/joerick/python-abi3-package-sample).

For the pyodide build it means the following change:

Previous wheel name: `zengl-2.3.0-cp311-cp311-emscripten_3_1_45_wasm32.whl`
New wheel name `zengl-2.3.0-cp311-abi3-emscripten_3_1_45_wasm32.whl`

And in theory when Pyodide will run on let say Python 3.13:

Previous wheel name: `zengl-2.3.0-cp313-cp313-emscripten_3_1_45_wasm32.whl`
New wheel name `zengl-2.3.0-cp311-abi3-emscripten_3_1_45_wasm32.whl`

The tag will always start with "cp311-abi3-".
Because, for this specific wheel the minimum limited api version is 3.11. Any python version 3.11+ can load this wheel.

My question is, does this have any unintended side effects on the pyodide side?

If so,

Should I patch this abi correction for the pyodide build?

Should I entirely disable building with the limited api set for pyodide?

I have tested and the currently it works fine, I cannot tell though what would happen when the python verison will not match the abi tag.

**Comment by hoodmane** [comment, marked as answer]
ABI3 works correctly with Pyodide. But it is useless until we stabilize the abi. We'll likely use a new Emscripten compiler when we update to python312 so instead of `cp312-cp312-emscripten_3_1_45_wasm32.whl` we'd have a platform of e.g., `cp312-cp312-emscripten_3_1_51_wasm32.whl`.

> Should I entirely disable building with the limited api set for pyodide?

I recommend this for now. We have been discussing ABI stabilization for a while but it's not clear when we'll bite the bullet on that. It's unfortunate for us that our compiler is both rapidly improving and has no ABI stability.

**Reply by szabolcsdombi** [reply]
I fully get it that the emscripten version must match so the previously built wheels are not usable.

If ABI3 works correctly it is very good news on my side. When building my library with Python3.12 and Emscriten 3.1.51 the tag will be `cp311-abi3-emscripten_3_1_51_wasm32.whl`.
